### PR TITLE
Cgal triangulation

### PR DESCRIPTION
**Pull request by luca-heltai**
Insertion of deal.II points into CGAL triangulation types, and convertion between CGAL Triangulation types to deal.II Triangulation types.

<s>Depends on #13644. Relevant changes to this PR can be seen here:

https://github.com/luca-heltai/dealii/compare/7cb4ee71..b1a8136e2
</s>

### CODE REVIEW

**Comment by luca-heltai** [comment]
image

This is the output of cgal_triangulation_05.cc

**Comment by luca-heltai** [comment]
Now that https://github.com/dealii/dealii/pull/13644 has been merged, this can be reviewed. This PR is holding back #13664 and https://github.com/luca-heltai/dealii/pull/111 (which will be opened right after  #13664 is merged).

File: doc/news/changes/major/20220426FederHeltai [hunk 1]
```diff
@@ -1,3 +1,12 @@
-New: Added support for the CGAL library (www.cgal.org).
+New: Added support for the CGAL library (www.cgal.org). The following features 
+are supported:
+<ul>
+  <li>Conversion from dealii to CGAL point types and viceversa</li>
```
Comment (peterrum) [inline doc/news/changes/major/20220426FederHeltai:4]:
```suggestion
  <li>Conversion from deal.II to CGAL point types and viceversa</li>
```

File: include/deal.II/cgal/triangulation.h [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+   * Both CGAL::Triangulation_2 and CGAL::Triangulation_3 can store degenerate
+   * two dimensional triangulations (i.e., a one-dimensional triangulation), or
+   * a degenerate three dimensional triangulation (i.e., a two-dimensional
+   * surface triangulation). The dimension of the triangulation, as returned by
+   * `cgal_triangulation.dimension()`, must match the dimension `dim` of the
+   * input Triangulation `dealii_triangulation`. If this is not the case, an
+   * exception is thrown.
+   *
+   * @param cgal_triangulation
+   * @param dealii_triangulation
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:99]: ?

File: include/deal.II/cgal/triangulation.h [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+#ifndef dealii_cgal_triangulation_h
+#define dealii_cgal_triangulation_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/grid/tria.h>
+
+#include <deal.II/cgal/utilities.h>
+
+#ifdef DEAL_II_WITH_CGAL
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:25]: Maybe move the ifdef above the includes?

File: include/deal.II/cgal/triangulation.h [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    Assert(triangulation.is_valid(),
+           ExcMessage(
+             "The triangulation you pass to this function should be a valid "
+             "CGAL triangulation."));
+    using CGALPoint = typename CGALTriangulation::Point;
+    std::vector<CGALPoint> cgal_points(points.size());
+    std::transform(
+      points.begin(), points.end(), cgal_points.begin(), [](const auto &p) {
+        return CGALWrappers::dealii_point_to_cgal_point<CGALPoint>(p);
+      });
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:125]:
```suggestion
    std::vector<CGALTriangulation::Point> cgal_points(points.size());
    std::transform(
      points.begin(), points.end(), cgal_points.begin(), [](const auto &p) {
        return CGALWrappers::dealii_point_to_cgal_point<CGALTriangulation::Point>(p);
      });
```

File: include/deal.II/cgal/triangulation.h [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                           ") does not match the dimension of the output "
+                           "deal.II triangulation (" +
+                           std::to_string(dim) + ")."));
+
+    Assert(dealii_triangulation.n_cells() == 0,
+           ExcMessage("The output triangulation object needs to be empty."));
+
+    const auto nv = cgal_triangulation.number_of_vertices();
+
+    // Deal.II storage data structures
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:154]:
```suggestion
    // deal.II storage data structures
```

File: include/deal.II/cgal/triangulation.h [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            CGALWrappers::cgal_point_to_dealii_point<spacedim>(v->point());
+          vertex_map[v] = i++;
+        }
+    }
+
+    auto has_faces = boost::hana::is_valid(
+      [](auto &&obj) -> decltype(obj.finite_face_handles()) {});
+
+    auto has_cells = boost::hana::is_valid(
+      [](auto &&obj) -> decltype(obj.finite_cell_handles()) {});
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:176]:
```suggestion
    const auto has_faces = boost::hana::is_valid(
      [](auto &&obj) -> decltype(obj.finite_face_handles()) {});

    const auto has_cells = boost::hana::is_valid(
      [](auto &&obj) -> decltype(obj.finite_cell_handles()) {});
```

File: include/deal.II/cgal/triangulation.h [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+    // Different loops for Triangulation_2 and Triangulation_3 types.
+    if constexpr (decltype(has_faces(cgal_triangulation)){})
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:179]: We cannot use `if constexp` yet.  That is only available once we switch to C++17. You will need to do template specialization.

File: include/deal.II/cgal/triangulation.h [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                           "deal.II triangulation (" +
+                           std::to_string(dim) + ")."));
+
+    Assert(dealii_triangulation.n_cells() == 0,
+           ExcMessage("The output triangulation object needs to be empty."));
+
+    const auto nv = cgal_triangulation.number_of_vertices();
+
+    // Deal.II storage data structures
+    std::vector<Point<spacedim>> vertices(nv);
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:155]:
```suggestion
    std::vector<Point<spacedim>> vertices(cgal_triangulation.number_of_vertices());
```

File: include/deal.II/cgal/triangulation.h [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                ExcMessage("The dimension of the input CGAL triangulation (" +
+                           std::to_string(cgal_triangulation.dimension()) +
+                           ") does not match the dimension of the output "
+                           "deal.II triangulation (" +
+                           std::to_string(dim) + ")."));
+
+    Assert(dealii_triangulation.n_cells() == 0,
+           ExcMessage("The output triangulation object needs to be empty."));
+
+    const auto nv = cgal_triangulation.number_of_vertices();
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:152]:
```suggestion
```

File: include/deal.II/cgal/triangulation.h [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    std::vector<Point<spacedim>> vertices(nv);
+    std::vector<CellData<dim>>   cells;
+    SubCellData                  subcell_data;
+
+    // CGAL storage data structures
+    std::map<typename CGALTriangulation::Vertex_handle, unsigned int>
+      vertex_map;
+    {
+      unsigned int i = 0;
+      for (auto v : cgal_triangulation.finite_vertex_handles())
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:164]:
```suggestion
      for (const auto & v : cgal_triangulation.finite_vertex_handles())
```
? Same below.

File: include/deal.II/cgal/triangulation.h [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+          for (const auto e : cgal_triangulation.finite_edges())
+            {
+              // An edge is idenfied by a face and a vertex index in the face
+              const auto &  f = e.first;
+              const auto &  i = e.second;
+              CellData<dim> cell(2);
+              unsigned int  id = 0;
+              for (unsigned int j = 0; j < 3; ++j)
+                if (j != i)
+                  cell.vertices[id++] = vertex_map[f->vertex(j)];
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:201]: What is the logic here?

File: include/deal.II/cgal/triangulation.h [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+      [](auto &&obj) -> decltype(obj.finite_cell_handles()) {});
+
+    // Different loops for Triangulation_2 and Triangulation_3 types.
+    if constexpr (decltype(has_faces(cgal_triangulation)){})
+      {
+        // This is a non-degenerate Triangulation_2
+        if (cgal_triangulation.dimension() == 2)
+          for (const auto f : cgal_triangulation.finite_face_handles())
+            {
+              CellData<dim> cell(3);
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:185]:
```suggestion
              CellData<dim> cell(ReferenceCells::Triangle.n_vertices());
```

File: include/deal.II/cgal/triangulation.h [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+                if (j != i)
+                  cell.vertices[id++] = vertex_map[c->vertex(j)];
+              cells.push_back(cell);
+            }
+        else if (cgal_triangulation.dimension() == 1)
+          // This is a degenerate Triangulation_3, made of edges
+          for (const auto edge : cgal_triangulation.finite_edges())
+            {
+              // An edge is idenfied by a cell and its two vertices
+              const auto &[c, i, j] = edge;
```
Comment (peterrum) [inline include/deal.II/cgal/triangulation.h:240]: This is also a C++17 feature, isn't it?

**Review by peterrum — COMMENTED** [review]
Overall looks good :+1:

File: include/deal.II/cgal/triangulation.h (hunk 13 above)
Comment (luca-heltai) [reply include/deal.II/cgal/triangulation.h:240]: Yes. The wrappers are only enabled if c++17 is enabled.

File: include/deal.II/cgal/triangulation.h (hunk 11 above)
Comment (luca-heltai) [reply include/deal.II/cgal/triangulation.h]: I've added a comment.

File: include/deal.II/cgal/triangulation.h (hunk 7 above)
Comment (luca-heltai) [reply include/deal.II/cgal/triangulation.h:181]: CGALWrappers are enabled only if c++17 is enabled.

**Comment by luca-heltai** [comment]
@peterrum, I've addressed all your comments.

File: include/deal.II/cgal/triangulation.h (hunk 7 above)
Comment (peterrum) [reply include/deal.II/cgal/triangulation.h:181]: OK. I see.

**Review by peterrum — APPROVED** [review]

**Comment by luca-heltai** [comment]
Failing is due to artifacts upload:

~~~
Unable to copy file to server StatusCode=BadRequest: Bad Request. Source file path: c:\project\dealii-windows.zip. Target server path: dealii-windows.zip/dealii-windows.zip
Cannot continue uploading files, so draining upload queue of 0 items.
Fail to upload 'c:\project\dealii-windows.zip' due to 'Critical failure uploading 'c:\project\dealii-windows.zip''.
GitHub.Runner.Plugins.Artifact.UploadFailedException: Critical failure uploading 'c:\project\dealii-windows.zip'
   at GitHub.Runner.Plugins.Artifact.FileContainerServer.UploadAsync(RunnerActionPluginExecutionContext context, Int32 uploaderId, CancellationToken token)
File error 'Critical failure uploading 'c:\project\dealii-windows.zip'' when uploading file 'c:\project\dealii-windows.zip'.
Associated artifact dealii-windows.zip (32189590) with run #66933
Error: Critical failure uploading 'c:\project\dealii-windows.zip'
Error: Exit code 1 returned from process: file name 'c:\runners\2.291.1\bin\Runner.PluginHost.exe', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'.
~~~

**Pull request merged**